cartographer/mapping_2d/sparse_pose_graph.cc

```diff
@@ -469,30 +469,37 @@ void SparsePoseGraph::RunOptimization() {
 
   const auto& submap_data = optimization_problem_.submap_data();
   const auto& node_data = optimization_problem_.node_data();
-  for (auto node_it = node_data.begin(); node_it != node_data.end();) {
-    const int trajectory_id = node_it->id.trajectory_id;
-    const auto trajectory_end = node_data.EndOfTrajectory(trajectory_id);
-    const int num_nodes = trajectory_nodes_.num_indices(trajectory_id);
-    for (; node_it != trajectory_end; ++node_it) {
-      const mapping::NodeId node_id = node_it->id;
-      auto& node = trajectory_nodes_.at(node_id);
-      node.pose =
-          transform::Embed3D(node_it->data.pose) *
-          transform::Rigid3d::Rotation(node.constant_data->gravity_alignment);
+
+  for (int trajectory_id : node_data.trajectory_ids()) {
+    auto last_optimized_node_id = mapping::INVALID_NODE_ID;
+    for (const auto& node : node_data.trajectory(trajectory_id)) {
+      auto& mutable_trajectory_node = trajectory_nodes_.at(node.id);
+      mutable_trajectory_node.pose =
+          transform::Embed3D(node.data.pose) *
+          transform::Rigid3d::Rotation(
+              mutable_trajectory_node.constant_data->gravity_alignment);
+      last_optimized_node_id = node.id;
```

Comment:
This will set `last_optimized_node_id` potentially many times. Why not
```
const mapping::NodeId last_optimized_node_id = std::prev(node_data.EndOfTrajectory(trajectory_id))->id;
```
after the loop similar to before? Would also remove the need for `INVALID_NODE_ID`.

Reply:
Done.